### ISSUE TITLE
[php] Unable to add file to DescriptorPool: couldn't resolve name '.google.protobuf.Timestamp' 

### ISSUE THREAD

**Issue by lazharichir** [google/protobuf#3966]
Hello - I have a Cloud Firestore project and am current doing some tests from my localhost to the FIrestore collections and documents. So far, I am trying to GET them and the very same query will sometimes work and sometimes throw an error:
```
PHP Fatal error:  Unable to add file to DescriptorPool: couldn't resolve name '.google.protobuf.Timestamp' in message 'google.firestore.v1beta1.TransactionOptions.ReadOnly'
 in /Users/Lazhar/myproject/services/article-service/v1/vendor/google/proto-client/src/GPBMetadata/Google/Firestore/V1Beta1/Common.php on line 43
```
I have no idea what is causing this occasionally but the very same query on the very same document in the very same collection will work one second later.

My composer.json is as follows:
```
{
    "name": "Article Service",
    "minimum-stability": "dev",
    "require": {
        "google/cloud": "0.*",
        "slim/slim": "^4.0",
        "vlucas/phpdotenv": "2.*"
    },
    "provide": {
        "psr/http-message-implementation": "1.0"
    },
    "autoload": {
        "psr-4": {
            "ArticleServiceAPI\\": "src",
            "ArticleService\\": "lib"
        }
    }
}
```
And the line 43 in the file in question is comment below:
```
        $pool->internalAddGeneratedFile(hex2bin(
            "0ac7050a25676f6f676c652f6669726573746f72652f763162657461312f" .
            "636f6d6d6f6e2e70726f746f1218676f6f676c652e6669726573746f7265" .
            "2e763162657461311a1f676f6f676c652f70726f746f6275662f74696d65" .
            "7374616d702e70726f746f22230a0c446f63756d656e744d61736b12130a" .
            "0b6669656c645f706174687318012003280922650a0c507265636f6e6469" .
            "74696f6e12100a06657869737473180120012808480012310a0b75706461" .
            "74655f74696d6518022001280b321a2e676f6f676c652e70726f746f6275" .
            "662e54696d657374616d70480042100a0e636f6e646974696f6e5f747970" .
            "6522b3020a125472616e73616374696f6e4f7074696f6e73124a0a097265" .
            "61645f6f6e6c7918022001280b32352e676f6f676c652e6669726573746f" .
            "72652e763162657461312e5472616e73616374696f6e4f7074696f6e732e" .
            "526561644f6e6c794800124c0a0a726561645f777269746518032001280b" .
            "32362e676f6f676c652e6669726573746f72652e763162657461312e5472" .
            "616e73616374696f6e4f7074696f6e732e52656164577269746548001a26" .
            "0a0952656164577269746512190a1172657472795f7472616e7361637469" .
            "6f6e18012001280c1a530a08526561644f6e6c79122f0a09726561645f74" .
            "696d6518022001280b321a2e676f6f676c652e70726f746f6275662e5469" .
            "6d657374616d70480042160a14636f6e73697374656e63795f73656c6563" .
            "746f7242060a046d6f64654298010a1c636f6d2e676f6f676c652e666972" .
            "6573746f72652e76316265746131420b436f6d6d6f6e50726f746f50015a" .
            "41676f6f676c652e676f6c616e672e6f72672f67656e70726f746f2f676f" .
            "6f676c65617069732f6669726573746f72652f763162657461313b666972" .
            "6573746f7265a2020447434653aa021e476f6f676c652e436c6f75642e46" .
            "69726573746f72652e56314265746131620670726f746f33" # this is line 43!
        ));
```

**Comment by TeBoring** [comment]
It will be very helpful if you can provide a small sample to reproduce the bug.
Are you using 3.5.0 release? php implementation of c extension? We recently provided well known types (including timestamp) in our runtime. Has filestore removed their own generated code for timestamp? Or have you updated both protobuf and filestore together?

**Comment by lazharichir** [comment]
I am using the 3.5.0 release and the official PHP extensions for gRPC and Protobuf from PECL, they both are up-to-date. GC Firestore library was also installed and updated from composer.

It is still doing it, re-updated everything:

```
[10-Dec-2017 11:48:58 Europe/Paris] PHP Fatal error:  Unable to add file to DescriptorPool: couldn't resolve name '.google.protobuf.Timestamp' in message 'google.firestore.v1beta1.TransactionOptions.ReadOnly'
 in /Users/Lazhar/topicseed/services/suggestions/vendor/google/proto-client/src/GPBMetadata/Google/Firestore/V1Beta1/Common.php on line 44
```
Example of this code:
```
	
	/**
	 * Firestore client
	 * @var \Google\Cloud\Firestore\FirestoreClient
	 */
	public static $db;

	/**
	 * Set the firestore client
	 * @return \Google\Cloud\Firestore\FirestoreClient $firestore Firestore client
	 */
	public static function firestore() {

		if( is_null(self::$db) ) {
			self::$db = new \Google\Cloud\Firestore\FirestoreClient([
				'projectId' => getenv('GC_PROJECT'),
				'keyFilePath' => getenv('GC_KEYFILE')
			]);
		}

		return self::$db;
	}

	public static function get( $seed ) {

		self::$seed = $seed;

		$start = microtime(1);

		$document = self::firestore()->collection('CachedSuggestions')->document($seed);

		if( $document->snapshot()->exists() ) {
			$diff = microtime(1)-$start;
			error_log('fetch firestore: ' . $diff);
			error_log('cache size: ' . $diff);
			return $document->snapshot()->fields();
		}

		error_log($seed . ' is NOT cached.');
		return;
	}

```

**Comment by TeBoring** [comment]
Sorry, I have difficulty reproducing your bug.
Could you paste your composer.json, command line code and your php code?

**Comment by TeBoring** [comment]
Please make sure you can use the stuff you paste to reproduce your bug.

**Comment by TeBoring** [comment]
Currently, I don't understand what is ```Example of this code:``` and how it's used to reproduce bug.

**Comment by TeBoring** [comment]
Make sure you to clean your previous ```vendor``` folder and do everything from scratch.

**Comment by miyurusagarage** [comment]
Hi, I'm encountering the same bug. It happens randomly on document create method.

![image](https://user-images.githubusercontent.com/5352806/33977917-461017a0-e0c3-11e7-8b4f-b75a5a888a40.png)

**Comment by lazharichir** [comment]
Here are two composer.json I have used @TeBoring :

```
{
    "name": "Organization Service",
    "description": "organizations",
    "minimum-stability": "dev",
    "require": {
        "slim/slim": "4.*",
        "google/cloud-firestore": "0.*",
        "vlucas/phpdotenv": "2.*"
    },
    "provide": {
        "psr/http-message-implementation": "1.0"
    },
    "autoload": {
        "psr-4": {
            "TS\\OrganizationService\\API\\": "src",
            "TS\\OrganizationService\\": "lib"
        }
    }
}
```
and... 
```
{
    "name": "User Service",
    "minimum-stability": "dev",
    "require": {
        "slim/slim": "4.*",
        "vlucas/phpdotenv": "2.*",
        "google/cloud-firestore": "^0.2.0"
    },
    "provide": {
        "psr/http-message-implementation": "1.0"
    },
    "autoload": {
        "psr-4": {
            "TS\\UserService\\API\\": "src",
            "TS\\UserService\\": "lib"
        }
    }
}
```
And a one-page index.php example I have just made that gives me the error, randomly, at reads:
```
<?php

// Autoload
require_once('../vendor/autoload.php');

// Load Enviroment Variables
$dotenv = new Dotenv\Dotenv('../');
$dotenv->load();

// Set the container's service for Firestore
$firestore = new \Google\Cloud\Firestore\FirestoreClient([
    'projectId' => getenv('GC_PROJECT'),
    'keyFilePath' => getenv('GC_PATH')
]);

$document = $firestore->collection('Organizations')->document('breedingbusiness');

if( ! $document->snapshot()->exists() ) {
    throw new \Exception("This organization does not exist.", 404);
} else {
    print_r($document->snapshot()->fields());
}
```
The error, again:
```
502 Bad Gateway
nginx/1.13.2

[14-Dec-2017 08:46:10 Europe/Madrid] PHP Fatal error:  Unable to add file to DescriptorPool: couldn't resolve name '.google.protobuf.Timestamp' in message 'google.firestore.v1beta1.TransactionOptions.ReadOnly'
 in /Users/Lazhar/ts/services/organization-service/vendor/google/proto-client/src/GPBMetadata/Google/Firestore/V1Beta1/Common.php on line 44
```

**Comment by TeBoring** [comment]
Here is my directory structure:
top:
composer.json vendor test/
test:
test.php (copy paste your index.php)

**Comment by TeBoring** [comment]
In test/, I ran: ```php test.php```
I got error:
```
Fatal error: Uncaught Dotenv\Exception\InvalidPathException: Unable to read the environment file at ../.env. in /usr/local/google/home/teboring/tmp/protobuf-git/php-bug/sandbox/vendor/vlucas/phpdotenv/src/Loader.php:98
Stack trace:
#0 /usr/local/google/home/teboring/tmp/protobuf-git/php-bug/sandbox/vendor/vlucas/phpdotenv/src/Loader.php(75): Dotenv\Loader->ensureFileIsReadable()
#1 /usr/local/google/home/teboring/tmp/protobuf-git/php-bug/sandbox/vendor/vlucas/phpdotenv/src/Dotenv.php(106): Dotenv\Loader->load()
#2 /usr/local/google/home/teboring/tmp/protobuf-git/php-bug/sandbox/vendor/vlucas/phpdotenv/src/Dotenv.php(50): Dotenv\Dotenv->loadData()
#3 /usr/local/google/home/teboring/tmp/protobuf-git/php-bug/sandbox/test/test.php(8): Dotenv\Dotenv->load()
#4 {main}
  thrown in /usr/local/google/home/teboring/tmp/protobuf-git/php-bug/sandbox/vendor/vlucas/phpdotenv/src/Loader.php on line 98
```

**Comment by TeBoring** [comment]
Please provide detailed steps to reproduce the bug. Otherwise there is few I can do.

**Comment by TeBoring** [comment]
Besides, the environment you are providing now have too much unrelated stuff. It really helps if you can minimize the environment and focus on the stuff that cause the bug.

**Comment by lazharichir** [comment]
Alright, so I setup a localhost server with nginx and php.

New top folder with the following under:
- index.php
- composer.json
- vendor/

![image](https://user-images.githubusercontent.com/12215372/34016530-85b2df0a-e122-11e7-86b3-48ce887a5b55.png)

Minimal contents of the two files index.php and composer.json:

INDEX.PHP
```
<?php

// Autoload
require_once('vendor/autoload.php');

// Set the container's service for Firestore
$firestore = new \Google\Cloud\Firestore\FirestoreClient([
    'projectId' => 'nameOfProject',
    'keyFilePath' => '/Users/Lazhar/Downloads/test/file_with_key.key'
]);

$document = $firestore->collection('Organizations')->document('breedingbusiness');

if( ! $document->snapshot()->exists() ) {
    throw new \Exception("This organization does not exist.", 404);
} else {
    print_r($document->snapshot()->fields());
}
```
COMPOSER.JSON
```
{
    "name": "github-test",
    "description": "",
    "minimum-stability": "dev",
    "require": {
        "google/cloud-firestore": "0.*"
    }
}
```
And at every few refreshes, I get the same error...
```
[14-Dec-2017 22:58:14 Europe/Madrid] PHP Fatal error:  Unable to add file to DescriptorPool: couldn't resolve name '.google.protobuf.Timestamp' in message 'google.firestore.v1beta1.TransactionOptions.ReadOnly'
 in /Users/Lazhar/Downloads/test/vendor/google/proto-client/src/GPBMetadata/Google/Firestore/V1Beta1/Common.php on line 44
```

**Comment by lazharichir** [comment]
It seems that running `php index.php` in terminal works so far the twenty times in a row i tried. The error often comes half the times if not more. I have just tried with Apache and it works fine, too.

So maybe it is linked to Nginx?

**Comment by TeBoring** [comment]
After running ```php index.php```, I got:
```
Fatal error: Uncaught Google\Cloud\Core\Exception\GoogleException: Given keyfile path /Users/Lazhar/Downloads/test/file_with_key.key does not exist in /usr/local/google/home/teboring/tmp/protobuf-git/php-bug/sandbox/vendor/google/cloud-core/ClientTrait.php:133
Stack trace:
#0 /usr/local/google/home/teboring/tmp/protobuf-git/php-bug/sandbox/vendor/google/cloud-core/ClientTrait.php(99): Google\Cloud\Firestore\FirestoreClient->getKeyFile(Array)
#1 /usr/local/google/home/teboring/tmp/protobuf-git/php-bug/sandbox/vendor/google/cloud-firestore/FirestoreClient.php(110): Google\Cloud\Firestore\FirestoreClient->configureAuthentication(Array)
#2 /usr/local/google/home/teboring/tmp/protobuf-git/php-bug/sandbox/test.php(8): Google\Cloud\Firestore\FirestoreClient->__construct(Array)
#3 {main}
  thrown in /usr/local/google/home/teboring/tmp/protobuf-git/php-bug/sandbox/vendor/google/cloud-core/ClientTrait.php on line 133
```

**Comment by TeBoring** [comment]
Could you reproduce your bug in a clean environment?

**Comment by TeBoring** [comment]
In order to run the test you provided, I need to have the file_with_key.key file.
Please either copy paste that file or change the test.

**Comment by lazharichir** [comment]
I cannot copy that file, it is my private json service key...

**Comment by TeBoring** [comment]
Do u have example how it should look like? Or give me a fake one.

**Comment by TeBoring** [comment]
If u can simply the test, it's even better.

**Comment by jackwakefield** [comment]
I'm having the same issue with:
`Unable to add file to DescriptorPool: couldn't resolve name '.google.protobuf.Timestamp' in message 'api.NotificationItem'`

`api.NotificationItem` being
```protobuf
message NotificationItem {
    int32 ID = 1;
    int32 CustomerID = 2;
    int32 ContractID = 3;
    string Type = 4;
    string Description = 5;
    google.protobuf.Timestamp Date = 6;
    bool Read = 7;
}
```

This is also producing the segfault:
```
Program received signal SIGSEGV, Segmentation fault.
message_free (object=0x7f62e6231bf0) at /tmp/pear/temp/protobuf/message.c:84
84	in /tmp/pear/temp/protobuf/message.c
(gdb) where
#0  message_free (object=0x7f62e6231bf0) at /tmp/pear/temp/protobuf/message.c:84
#1  0x00007f62f2711e91 in zend_objects_store_free_object_storage () from /usr/lib/apache2/modules/libphp7.so
#2  0x00007f62f26c2323 in shutdown_executor () from /usr/lib/apache2/modules/libphp7.so
#3  0x00007f62f26d361b in zend_deactivate () from /usr/lib/apache2/modules/libphp7.so
#4  0x00007f62f266e8c2 in php_request_shutdown () from /usr/lib/apache2/modules/libphp7.so
#5  0x00007f62f277aed7 in php_handler () from /usr/lib/apache2/modules/libphp7.so
#6  0x000055f4f7c14880 in ap_run_handler ()
#7  0x000055f4f7c14dc9 in ap_invoke_handler ()
#8  0x000055f4f7c2a5cc in ap_internal_redirect ()
#9  0x00007f62ed2d0ea2 in ?? () from /usr/lib/apache2/modules/mod_rewrite.so
#10 0x000055f4f7c14880 in ap_run_handler ()
#11 0x000055f4f7c14dc9 in ap_invoke_handler ()
#12 0x000055f4f7c2aca2 in ap_process_async_request ()
#13 0x000055f4f7c2ae40 in ap_process_request ()
#14 0x000055f4f7c27742 in ?? ()
#15 0x000055f4f7c1e130 in ap_run_process_connection ()
#16 0x00007f62f31fb7ba in ?? () from /usr/lib/apache2/modules/mod_mpm_prefork.so
#17 0x00007f62f31fba01 in ?? () from /usr/lib/apache2/modules/mod_mpm_prefork.so
#18 0x00007f62f31fba72 in ?? () from /usr/lib/apache2/modules/mod_mpm_prefork.so
#19 0x00007f62f31fc73c in ?? () from /usr/lib/apache2/modules/mod_mpm_prefork.so
#20 0x000055f4f7bf884e in ap_run_mpm ()
#21 0x000055f4f7bf1673 in main ()
```

**Comment by TeBoring** [comment]
Fixed in #4139
